### PR TITLE
AF-105: Fix Upload Example Crash

### PR DESCRIPTION
**Pull request by supperspidey**
#### Ticket

[AF-105](https://vimean.atlassian.net/browse/AF-105)

#### Pull Request Checklist

- [x] Resolved any merge conflicts
- [x] No build errors or warnings are introduced
- [x] New files are written in Swift
- [x] New classes contain license headers
- [x] New classes have Documentation
- [x] New public methods have Documentation

#### Issue Summary

This PR fixes the crashes occurring while attempting to upload a video via the sample app. Originally, the crash was reported for `VimeoUpload` target, which implements the two-step upload flow. Since two-step upload hasn't been available to the public yet, it was found that any current attempt to upload via this flow would not have the video object returned. That behavior caused the crash that multiple people have reported because the flow assumed the video object was always there. The change in the `VimeoUpload` target addresses this issue by safely unwrapping the video object &mdash; if it does not exist, then we show an alert to the user and not execute the remaining logic.

In addition, it was discovered that there was another bug in the four-step upload flow that caused a crash when attempting to upload. The crash occurred because the run-time did not recognize the `encode(with aCoder: NSCoder)` method from the `VimeoUpload` class. This was a direct result from the Swift 3 conversion. This PR also includes a fix to this issue.

#### Implementation Summary

- For the crash in the two-step flow, the video object is now safely unwrapped using `guard let`.
- For the crash in the four-step flow, `VimeoUpload` now conforms to `NSCoding` protocol. Because of that, `encode(with aCoder: NSCoder)` is now `public`, thus making it recognizable to the runtime.

#### Reviewer Tips

For the four-step upload flow, make sure you use a token that has `edit` and `upload` scope.

#### How to Test

- Make sure the two-step upload flow is not available to any third-party access token.
- Make sure the four-step upload flow is able to upload a video with an access token whose scopes include `upload` and `edit`.

### CODE REVIEW

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          6 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/Upload/Descriptor System/Old Upload/OldUploadDescriptor.swift#L50">VimeoUpload/Upload/Descriptor System/Old Upload/OldUploadDescriptor.swift#L50</a></strong>: string interpolation produces a debug description for an optional value; did you mean to make this explicit?  <br /><code>
            print("\(self.currentRequest.rawValue) \(self.identifier)")
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/UIKit/DescriptorKVObserver.swift#L31">VimeoUpload/UIKit/DescriptorKVObserver.swift#L31</a></strong>: extraneous duplicate parameter name; ‘descriptorType’ already has an argument label  <br /><code>
    func descriptorStateDidChange(descriptorType descriptorType: VideoDescriptorType, descriptorState: DescriptorState, isCancelled: Bool, descriptorError: NSError?)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/UIKit/DescriptorKVObserver.swift#L32">VimeoUpload/UIKit/DescriptorKVObserver.swift#L32</a></strong>: extraneous duplicate parameter name; ‘descriptorType’ already has an argument label  <br /><code>
    func descriptorProgressDidChange(descriptorType descriptorType: VideoDescriptorType, progress: Double)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/Descriptor System/DescriptorManager.swift#L232">VimeoUpload/Descriptor System/DescriptorManager.swift#L232</a></strong>: conditional downcast from ‘Error?’ to ‘NSError’ is a bridging conversion; did you mean to use ‘as’?  <br /><code>
                let isConnectionError = ((task.error as? NSError)?.isConnectionError() == true || (error as? NSError)?.isConnectionError() == true)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32">VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32</a></strong>: extraneous duplicate parameter name; ‘completion’ already has an argument label  <br /><code>
    func approximateFileSize(completion completion: @escaping FloatBlock)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/ee33cbcb6aaf5bd816228c46d6fd737e5a8375c9/VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift#L27">VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift#L27</a></strong>: string interpolation produces a debug description for an optional value; did you mean to make this explicit?  <br /><code>
            print("\(self.currentRequest.rawValue) \(self.identifier)")
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.309 (0.310) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.330 (0.332) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoUpload/193)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by supperspidey** [comment]
cc @nicolelehrer

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift [hunk 1]
```diff
@@ -285,8 +285,20 @@ class VideoSettingsViewController: UIViewController, UITextFieldDelegate
     
     private func applyVideoSettings()
     {
-        let videoUri = self.uploadTicket!.video!.uri!
-        let videoSettings = self.videoSettings!
+        guard let videoUri = self.uploadTicket?.video?.uri, let videoSettings = self.videoSettings else
+        {
+            let alertController = UIAlertController(
+                title: "Cannot Upload Video",
```
Comment (jasonhawkins) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift]: I know the rest of the file doesn't do this yet, but can we move these strings into constants? Or a struct for constants?

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift [hunk 2]
```diff
@@ -285,8 +285,20 @@ class VideoSettingsViewController: UIViewController, UITextFieldDelegate
     
     private func applyVideoSettings()
     {
-        let videoUri = self.uploadTicket!.video!.uri!
-        let videoSettings = self.videoSettings!
+        guard let videoUri = self.uploadTicket?.video?.uri, let videoSettings = self.videoSettings else
```
Comment (jasonhawkins) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift]: Pretty minor, but can we use `videoURI` casing instead?

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift [hunk 3]
```diff
@@ -285,8 +285,20 @@ class VideoSettingsViewController: UIViewController, UITextFieldDelegate
...
+        {
+            let alertController = UIAlertController(
+                title: "Cannot Upload Video",
+                message: "The video object in `uploadTicket` does not exist!",
```
Comment (nicolelehrer) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift]: Should this somehow tell the user that if they are trying to upload using new-upload target (which uses 4 step upload), it won't work? @supperspidey

Comment (supperspidey) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift]: You mean 2-step @nicolelehrer ? It's actually a good idea.

**Comment by supperspidey** [comment]
Thanks @nicolelehrer and @jasonhawkins !

**Pull request merged**